src/libraries/Microsoft.PowerFx.Core/Public/Types/TypeCheckers/AggregateTypeChecker.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Public.Types.TypeCheckers
+{
+    internal abstract class AggregateTypeChecker
+    {
+        protected ICollection<ExpressionError> _errorList;
+
+        public AggregateTypeChecker(ICollection<ExpressionError> errorList)
+        {
+            _errorList = errorList ?? throw new ArgumentNullException(nameof(errorList));
+        }
+
+        public bool Run(AggregateType sourceType, AggregateType typeToCheck)
+        {
+            if (sourceType == null || typeToCheck == null)
+            {
+                return false;
+            }
+            else if (sourceType._type.Kind != typeToCheck._type.Kind)
+            {
+                _errorList.Add(new ExpressionError()
+                {
+                    Kind = ErrorKind.Validation,
+                    Severity = ErrorSeverity.Critical,
+                    Message = $"Type mismatch between source and target types. Expected: {sourceType._type.GetKindString()}, found {typeToCheck._type.GetKindString()}."
+                });
+                return false;
+            }
+
+            return IsMatch(sourceType, typeToCheck);
+        }
+
+        public abstract bool IsMatch(AggregateType sourceType, AggregateType typeToCheck);
+
+        public abstract bool IsFieldMatch(FormulaType sourceField, FormulaType fieldToCheck);
+    }
+
+    internal class StrictAggregateTypeChecker : AggregateTypeChecker
+    {
+        public StrictAggregateTypeChecker(ICollection<ExpressionError> errorList)
+            : base(errorList)
+        {
+        }
+
+        public override bool IsMatch(AggregateType sourceType, AggregateType typeToCheck)
+        {
+            var sourceFieldCount = sourceType.FieldNames.Count();
+            var typeToCheckFieldCount = typeToCheck.FieldNames.Count();
```

Comment:
This could be a problem in expansion (dataverse record)  case. 

can we check TableSymbolName, and do something like this before we compare fields:

var x =source.TableSymbolName;
var y = typeToChecl.TableSymbolName;

if (x != null || y != null) {
  if (x != y) {
    // error 
  }
}

Reply:
Added the check.